src/dynamicalsystem.jl

```diff
@@ -0,0 +1,349 @@
+using OrdinaryDiffEq, ForwardDiff, StaticArrays
+import OrdinaryDiffEq: isinplace, ODEIntegrator, step!
+import Base: eltype
+
+export dimension, state, DynamicalSystem, DS, integrator, tangent_integrator
+export ContinuousDynamicalSystem, CDS, DiscreteDynamicalSystem, DDS
+export set_parameter!, step!
+
+#######################################################################################
+#                          Basic functions and interface                              #
+#######################################################################################
+
+dimension(prob::DEProblem) = length(prob.u0)
+eltype(prob::DEProblem) = eltype(prob.u0)
+state(prob::DEProblem) = prob.u0
+hascallback(prob::DEProblem) = prob.callback != nothing
+statetype(prob::DEProblem) = eltype(prob.u0)
+
+systemtype(::ODEProblem) = "continuous"
+systemtype(::DiscreteProblem) = "discrete"
+
+function extract_solver(diff_eq_kwargs)
```

Comment:
Isn't it equivalent to this?

```julia
function extract_solver(diff_eq_kwargs)
    newkw = deepcopy(diff_eq_kwargs)
    solver = pop!(newkw, DEFAULT_SOLVER)
    return solver, newkw
end
```

I mean, assuming this is a small dictionary so that deepcopy is super cheap.